File: python/src/algo/sort/heap.py
import random

def left(i):
    return 2*i + 1
def right(i):
    return 2*i + 2
def parent(i):
    return (i-1) // 2
def is_leaf(aList, i):
    return 2 * i >= len(aList)

def heapify_node(L, llen, node):
    # assume the child nodes are both heaps already
    l = left(node)
    if l < llen and L[l] > L[node]:
        largest = l
    else:
        largest = node

    r = right(node)
    if r < llen and L[r] > L[largest]:
        largest = r

    if largest != node:
        L[largest], L[node] = L[node], L[largest]
        heapify_node(L, llen, largest)


def heapify(L, llen):
    for i in range((llen-1)//2, -1, -1):
        heapify_node(L, llen, i)
    print(L[:llen])

def heap_sort(L):
    heapify(L, len(L))
    for i in range(len(L)-1, 0, -1):
        L[0], L[i] = L[i], L[0]
        heapify_node(L, i, 0)

if __name__ == "__main__":
    L = list(range(20))
    random.shuffle(L)
    print(L)
    # heapify(L, len(L))
    heap_sort(L)
    print(L)

File: cpp/algo/heap.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <random>
using namespace std;

template<class T>
class heap{
private:
	vector<T> &A;
public:
	heap(vector<T> &v) : A(v) {
		cout << "in heap" << endl;
	}
	int left(int n){ return 2*n + 1; }
	int right(int n){ return 2*n + 2; }
	int parent(int n){ return n > 0 ? (n-1)/2 : 0; }

	// Assume that the binary trees rooted at LEFT(i) and RIGHT(i)
	// are max-heaps, but A[i] might be smaller than its children
	// and thus violating the heap property
	// Time complexity: O(lgn)
	void max_heapify(int i, int heap_size){
	    int l = left(i);
	    int r = right(i);
	    int largest;
	    if (l < heap_size && A[l] > A[i])
	        largest = l;
	    else
	        largest = i;

	    if (r < heap_size && A[r] > A[largest])
	        largest = r;

	    if (largest != i){
	        swap(A[largest], A[i]);
	        max_heapify(largest, heap_size);
        }
	}

    // Time complexity: O(n)
    // Note here that O(nlgn) is not tight
    void build_max_heap(int heap_size){
        // Note leaves are already heapified
        for (int i=parent(A.size()-1); i>=0; --i){
            max_heapify(i, heap_size);
        }
    }

    void heapsort(){
        build_max_heap(A.size());
        for (int i=0; i<A.size(); ++i){
            swap(A[0], A[A.size()-1-i]);
            max_heapify(0, A.size()-1-i);
        }
    }


};

int main(int argc, char *argv[])
{
    // First create an instance of an engine.
    random_device rnd_device;

    // Specify the engine and distribution.
    mt19937 mersenne_engine {rnd_device()};  // Generates random integers
    uniform_int_distribution<int> dist {1, 52};

    auto gen = [&dist, &mersenne_engine](){
                   return dist(mersenne_engine);
               };

    vector<int> vec(10);
    generate(begin(vec), end(vec), gen);

    copy(vec.begin(), vec.end(), ostream_iterator<int>(cout, " "));
    cout << endl;

    (heap<int>(vec)).heapsort();

    copy(vec.begin(), vec.end(), ostream_iterator<int>(cout, " "));
    cout << endl;

}
